docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+$ jobserver --address=127.0.0.1:9090 --cacert=ca.crt --cert=tls.crt --key=tls.key ...
+
+# starting a job:
+# for commands that don't require flag args, they can be passed as-is:
+$ jobserver run kubectl logs pod/example
+<job-id> # job id is printed
+
+# for commands that do require flag args, the expected '--' delimiter can be used:
+$ jobserver run -- kubectl logs --follow pod/example
+<job-id>
+
+# to start a job with resource limits:
+$ jobserver run \
+  --cpus=100m \ # 100 milli-cores; can also write e.g. '--cpus=2' for 2 cores (no float values)
+  --memory=1Gi --memory-soft-limit=512Mi \ # the familiar suffixes can be used for memory limits
+  --device-read-bps=8:16=200 \ # format is major:minor=value (bytes per second)
+  --device-write-bps=8:16=200,8:0=50 \ # multiple devices can be specified by separating with commas
+  --device-read-iops=... --device-write-iops=... \ # same format, but units are in iops
+  -- go build -o bin/jobserver ./cmd/jobserver
+
+# stopping a job:
+$ jobserver stop <job-id>
+<job-id>
+
+# getting the status of a job:
+$ jobserver status <job-id>
+state: running
+pid: 1234
+start time: 2006-01-02T15:04:05Z07:00
+
+# getting the output of a job:
+$ jobserver logs [--follow] <job-id>
+<output>
+
+# getting a list of all job ids:
+$ jobserver list
+JOB ID      COMMAND       CREATED          STATUS
+<id-1>      kubectl       2006-01-02...    Running
+<id-2>      go            2006-01-02...    Completed
+```
+
+### Implementation Details
+
+#### Streaming Output
+
+Output streaming will be implemented as follows:
+
+1. When a job is started, we will pipe the combined stdout and stderr of its process to a simple in-memory buffer.
+2. When a client initiates a stream by calling `Output()`, we will write the current contents of the buffer to the stream in 4MB chunks. Then, if the client requested to follow the output, the server will keep the stream open, and any subsequent reads from the process's output will be duplicated to all clients that are following the stream, in addition to the server's own buffer.
+
+#### Process Lifecycle
+
+For jobs that do not specify resource limits, they will be run as a child process of the job server itself without any additional isolation. For jobs that do specify resource limits, they will be run in their own cgroup with all of the requested limits.
+
+On startup, the server will do the following (simplifying, there will be additional abstractions here):
+
+1. Ensure its euid is 0 so it can manage cgroups, and ensure that cgroups v2 is enabled.
+2. Create a parent cgroup in `/user.slice/user-<uid>.slice/jobserver` if it doesn't exist. The specific path can be determined by first identifying its own cgroup from `/proc/self/cgroup` (which, if run as `sudo jobserver serve`, will be something like `/user.slice/user-<uid>.slice/<terminal emulator's cgroup>/vte-spawn-<some uuid>.scope`), then walking the tree up until it finds a cgroup matching `user-*.slice`.
```

Comment:
Since we seem to be doing systemd things I can't help but mention that `systemd-run --user --scope` will execute a process in a dedicated systemd scope (unmanaged cgroup), and you can configure systemd to delegate cgroup management to your specific user, so you don't actually need to be uid 0 for this.

I'd just work inside the cgroup that the process is spawned in rather than get locked into systemd-isms -- or just use the hierarchy rooted at `/kralicky-jobserver-<nonce>/` instead, if you require superuser privileges anyway.

Reply:
Good points. I thought about using the server's own cgroup as the parent but it seems like (at least on my system, not sure if this is always the case?) the only controllers enabled in it are memory and pids, all the way up to `/user@<uid>.service` which adds cpuset, cpu, and io.

I think the top level cgroup idea is good, it would be a simple solution and works fine for the purposes of this challenge :+1: